libcnb-test/src/container_port_mapping.rs

```diff
@@ -0,0 +1,228 @@
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) enum PortMapParseError {
+    UnexpectedOrMissingContainerPortSuffix,
+    ContainerPortParseError(ParseIntError),
+    NoBindingForPort(u16),
+    MissingHostIpAddress,
+    MissingHostPort,
+    HostIpAddressParseError(AddrParseError),
+    HostPortParseError(ParseIntError),
+}
+
+fn parse_port_map_entry(
+    port_mapping: (&String, &Option<Vec<PortBinding>>),
+) -> Result<(u16, SocketAddr), PortMapParseError> {
+    let (container_port, mappings) = port_mapping;
+
+    let container_port = container_port
+        .strip_suffix("/tcp")
+        .ok_or(PortMapParseError::UnexpectedOrMissingContainerPortSuffix)
+        .and_then(|port_string| {
+            port_string
+                .parse()
+                .map_err(PortMapParseError::ContainerPortParseError)
+        })?;
+
+    let port_binding = mappings
+        .clone()
+        .unwrap_or_default()
+        .first()
+        .cloned()
+        .ok_or(PortMapParseError::NoBindingForPort(container_port))?;
+
+    let host_address = port_binding
+        .host_ip
+        .ok_or(PortMapParseError::MissingHostIpAddress)
+        .and_then(|host_ip| {
+            host_ip
+                .parse()
+                .map_err(PortMapParseError::HostIpAddressParseError)
+        })?;
+
+    let host_port = port_binding
+        .host_port
+        .ok_or(PortMapParseError::MissingHostPort)
+        .and_then(|host_port| {
+            host_port
+                .parse()
+                .map_err(PortMapParseError::HostPortParseError)
+        })?;
+
+    Ok((container_port, SocketAddr::new(host_address, host_port)))
+}
+
+/// Create a new Bollard container config with the given exposed ports.
+///
+/// The exposed ports will be forwarded to random ports on the host.
+pub(crate) fn port_mapped_container_config(ports: &[u16]) -> bollard::container::Config<String> {
+    Config {
+        host_config: Some(HostConfig {
+            port_bindings: Some(
+                ports
+                    .iter()
+                    .map(|port| {
+                        (
+                            format!("{}/tcp", port),
+                            Some(vec![PortBinding {
+                                host_ip: None,
+                                host_port: None,
+                            }]),
+                        )
+                    })
+                    .collect::<PortMap>(),
+            ),
+            ..Default::default()
+        }),
+        exposed_ports: Some(
+            ports
+                .iter()
+                .map(|port| {
+                    (
+                        format!("{}/tcp", port),
+                        #[allow(clippy::zero_sized_map_values)]
+                        HashMap::new(), // Bollard requires this zero sized value map,
+                    )
+                })
+                .collect(),
+        ),
+        ..Default::default()
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
There aren't tests for most of the error cases in `parse_port_map_entry()`. I'm presuming this is intentional, but wanted to mention in case not.

Reply:
Yeah intentional, that would be a lot of test code for cases that I've never seen occurring.